File: pkg/datasource/stock.go
package datasource

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type StockDataSource struct {
	symbol          string
	apiKey          string
	updateInterval  time.Duration
	currentData     StockData
	subscribers     []func(value interface{})
	ticker          *time.Ticker
	isRunning       bool
}

// AlphaVantageResponse represents the API response structure
type AlphaVantageResponse struct {
	GlobalQuote struct {
		Symbol           string `json:"01. symbol"`
		Open             string `json:"02. open"`
		High             string `json:"03. high"`
		Low              string `json:"04. low"`
		Price            string `json:"05. price"`
		Volume           string `json:"06. volume"`
		LatestTradingDay string `json:"07. latest trading day"`
		PreviousClose    string `json:"08. previous close"`
		Change           string `json:"09. change"`
		ChangePercent    string `json:"10. change percent"`
	} `json:"Global Quote"`
}

// FinnhubResponse represents Finnhub API response
type FinnhubResponse struct {
	C  float64 `json:"c"`  // Current price
	D  float64 `json:"d"`  // Change
	DP float64 `json:"dp"` // Percent change
	H  float64 `json:"h"`  // High price of the day
	L  float64 `json:"l"`  // Low price of the day
	O  float64 `json:"o"`  // Open price of the day
	PC float64 `json:"pc"` // Previous close price
}

func NewStockDataSource(symbol string, apiKey string) *StockDataSource {
	return &StockDataSource{
		symbol:         strings.ToUpper(symbol),
		apiKey:         apiKey,
		updateInterval: 5 * time.Minute, // Update every 5 minutes to avoid rate limits
		subscribers:    make([]func(value interface{}), 0),
		currentData: StockData{
			Symbol:      strings.ToUpper(symbol),
			DisplayText: fmt.Sprintf("📈 %s: Loading...", strings.ToUpper(symbol)),
		},
	}
}

func (s *StockDataSource) Start(ctx context.Context) {
	if s.isRunning {
		return
	}
	s.isRunning = true

	// Initial fetch
	s.fetchStockData()

	// Start periodic updates
	s.ticker = time.NewTicker(s.updateInterval)
	go func() {
		for {
			select {
			case <-ctx.Done():
				if s.ticker != nil {
					s.ticker.Stop()
				}
				s.isRunning = false
				return
			case <-s.ticker.C:
				s.fetchStockData()
			}
		}
	}()
}

func (s *StockDataSource) Subscribe(callback func(value interface{})) {
	s.subscribers = append(s.subscribers, callback)
}

func (s *StockDataSource) GetCurrentValue() interface{} {
	return s.currentData
}

func (s *StockDataSource) fetchStockData() {
	fmt.Printf("Fetching stock data for %s...\n", s.symbol)
	
	// Try free Yahoo Finance API first
	data, err := s.fetchFromYahooFinance()
	if err != nil {
		fmt.Printf("Yahoo Finance error: %v, trying Finnhub...\n", err)
		
		// Try Finnhub if API key is available
		data, err = s.fetchFromFinnhub()
		if err != nil {
			fmt.Printf("Finnhub error: %v, trying Alpha Vantage...\n", err)
			// Fallback to Alpha Vantage if available
			data, err = s.fetchFromAlphaVantage()
			if err != nil {
				fmt.Printf("Alpha Vantage error: %v\n", err)
				s.currentData = StockData{
					Symbol:      s.symbol,
					DisplayText: fmt.Sprintf("📈 %s: Error fetching data", s.symbol),
					PriceString: "---",
					ChangeString: "---",
				}
			} else {
				s.currentData = *data
			}
		} else {
			s.currentData = *data
		}
	} else {
		s.currentData = *data
	}

	fmt.Printf("Stock data updated: %s\n", s.currentData.DisplayText)
	
	// Notify subscribers
	for _, callback := range s.subscribers {
		callback(s.currentData)
	}
}

func (s *StockDataSource) fetchFromFinnhub() (*StockData, error) {
	// Finnhub requires a real API key now
	if s.apiKey == "" {
		return nil, fmt.Errorf("Finnhub API key required")
	}
	url := fmt.Sprintf("https://finnhub.io/api/v1/quote?symbol=%s&token=%s", s.symbol, s.apiKey)
	
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var finnhubData FinnhubResponse
	if err := json.Unmarshal(body, &finnhubData); err != nil {
		return nil, err
	}

	// Check if we got valid data
	if finnhubData.C == 0 {
		return nil, fmt.Errorf("no data available for symbol %s", s.symbol)
	}

	return s.formatStockData(finnhubData.C, finnhubData.D, finnhubData.DP), nil
}

func (s *StockDataSource) fetchFromAlphaVantage() (*StockData, error) {
	if s.apiKey == "" {
		return nil, fmt.Errorf("Alpha Vantage API key not provided")
	}
	
	url := fmt.Sprintf("https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=%s&apikey=%s", s.symbol, s.apiKey)
	
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var alphaData AlphaVantageResponse
	if err := json.Unmarshal(body, &alphaData); err != nil {
		return nil, err
	}

	if alphaData.GlobalQuote.Symbol == "" {
		return nil, fmt.Errorf("no data available for symbol %s", s.symbol)
	}

	price, _ := strconv.ParseFloat(alphaData.GlobalQuote.Price, 64)
	change, _ := strconv.ParseFloat(alphaData.GlobalQuote.Change, 64)
	
	// Parse change percent (remove % sign)
	changePercentStr := strings.TrimSuffix(alphaData.GlobalQuote.ChangePercent, "%")
	changePercent, _ := strconv.ParseFloat(changePercentStr, 64)

	return s.formatStockData(price, change, changePercent), nil
}

func (s *StockDataSource) formatStockData(price, change, changePercent float64) *StockData {
	isPositive := change >= 0
	isNegative := change < 0
	
	var trendEmoji string
	var changeColor string
	
	if isPositive {
		trendEmoji = "📈"
		changeColor = "+"
	} else {
		trendEmoji = "📉"
		changeColor = ""
	}

	priceStr := fmt.Sprintf("$%.2f", price)
	changeStr := fmt.Sprintf("%s%.2f (%.2f%%)", changeColor, change, changePercent)
	
	displayText := fmt.Sprintf("%s %s: %s %s", trendEmoji, s.symbol, priceStr, changeStr)
	
	return &StockData{
		Symbol:        s.symbol,
		Price:         price,
		Change:        change,
		ChangePercent: changePercent,
		DisplayText:   displayText,
		PriceString:   priceStr,
		ChangeString:  changeStr,
		IsPositive:    isPositive,
		IsNegative:    isNegative,
		TrendEmoji:    trendEmoji,
	}
}

func (s *StockDataSource) fetchFromYahooFinance() (*StockData, error) {
	// Use Yahoo Finance query API (free, no key required)
	url := fmt.Sprintf("https://query1.finance.yahoo.com/v8/finance/chart/%s", s.symbol)
	
	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	
	// Add headers to look like a regular browser request
	req.Header.Set("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36")
	req.Header.Set("Accept", "application/json")
	
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var yahooData struct {
		Chart struct {
			Result []struct {
				Meta struct {
					Currency                string  `json:"currency"`
					Symbol                  string  `json:"symbol"`
					RegularMarketPrice      float64 `json:"regularMarketPrice"`
					PreviousClose           float64 `json:"previousClose"`
					ChartPreviousClose      float64 `json:"chartPreviousClose"`
				} `json:"meta"`
			} `json:"result"`
		} `json:"chart"`
	}

	if err := json.Unmarshal(body, &yahooData); err != nil {
		return nil, err
	}

	if len(yahooData.Chart.Result) == 0 {
		return nil, fmt.Errorf("no data available for symbol %s", s.symbol)
	}

	meta := yahooData.Chart.Result[0].Meta
	if meta.RegularMarketPrice == 0 {
		return nil, fmt.Errorf("invalid price data for symbol %s", s.symbol)
	}

	price := meta.RegularMarketPrice
	previousClose := meta.PreviousClose
	if previousClose == 0 {
		previousClose = meta.ChartPreviousClose
	}
	
	change := price - previousClose
	changePercent := 0.0
	if previousClose != 0 {
		changePercent = (change / previousClose) * 100
	}

	return s.formatStockData(price, change, changePercent), nil
}

// Convenience function to create NASDAQ stock data source
func NewNASDAQStockDataSource(symbol string, apiKey string) *StockDataSource {
	return NewStockDataSource(symbol, apiKey)
}

File: rust/src/sources/stock.rs
//! Stock quote. Port of pkg/datasource/stock.go.
//! Tries Yahoo Finance (no key), then Finnhub, then Alpha Vantage (need a key).

use std::time::Duration as StdDuration;

use serde_json::Value;

#[allow(dead_code)] // full quote fields mirror the Go model
#[derive(Debug, Clone, Default)]
pub struct StockData {
    pub symbol: String,
    pub price: f64,
    pub change: f64,
    pub change_percent: f64,
    pub display_text: String,
    pub price_string: String,
    pub change_string: String,
    pub is_positive: bool,
    pub is_negative: bool,
    pub trend_emoji: String,
}

pub fn config() -> (String, String) {
    let symbol = std::env::var("EZBAR_STOCK_SYMBOL")
        .ok()
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| "NQ=F".to_string())
        .to_uppercase();
    let api_key = std::env::var("EZBAR_STOCK_API_KEY").unwrap_or_default();
    (symbol, api_key)
}

fn format_stock(symbol: &str, price: f64, change: f64, change_percent: f64) -> StockData {
    let is_positive = change >= 0.0;
    let (trend_emoji, sign) = if is_positive { ("📈", "+") } else { ("📉", "") };
    let price_string = format!("${:.2}", price);
    let change_string = format!("{}{:.2} ({:.2}%)", sign, change, change_percent);
    let display_text = format!("{} {}: {} {}", trend_emoji, symbol, price_string, change_string);
    StockData {
        symbol: symbol.to_string(),
        price,
        change,
        change_percent,
        display_text,
        price_string,
        change_string,
        is_positive,
        is_negative: change < 0.0,
        trend_emoji: trend_emoji.to_string(),
    }
}

fn http() -> Result<reqwest::Client, String> {
    reqwest::Client::builder()
        .timeout(StdDuration::from_secs(15))
        .build()
        .map_err(|e| e.to_string())
}

pub async fn fetch(symbol: &str, api_key: &str) -> Result<StockData, String> {
    match fetch_yahoo(symbol).await {
        Ok(d) => return Ok(d),
        Err(e) => log::warn!("yahoo: {e}"),
    }
    if !api_key.is_empty() {
        match fetch_finnhub(symbol, api_key).await {
            Ok(d) => return Ok(d),
            Err(e) => log::warn!("finnhub: {e}"),
        }
        match fetch_alphavantage(symbol, api_key).await {
            Ok(d) => return Ok(d),
            Err(e) => log::warn!("alphavantage: {e}"),
        }
    }
    Ok(StockData {
        symbol: symbol.to_string(),
        display_text: format!("📈 {}: Error fetching data", symbol),
        price_string: "---".to_string(),
        change_string: "---".to_string(),
        ..Default::default()
    })
}

/// Fetches 7 days of hourly closes for the hover chart (Yahoo Finance).
pub async fn fetch_chart(symbol: &str) -> Vec<f64> {
    let url =
        format!("https://query1.finance.yahoo.com/v8/finance/chart/{symbol}?interval=1h&range=7d");
    let client = match http() {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };
    let resp = match client
        .get(&url)
        .header("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36")
        .header("Accept", "application/json")
        .send()
        .await
    {
        Ok(r) => r,
        Err(_) => return Vec::new(),
    };
    let v: Value = match resp.json().await {
        Ok(v) => v,
        Err(_) => return Vec::new(),
    };
    v["chart"]["result"][0]["indicators"]["quote"][0]["close"]
        .as_array()
        .map(|arr| arr.iter().filter_map(|c| c.as_f64()).collect())
        .unwrap_or_default()
}

async fn fetch_yahoo(symbol: &str) -> Result<StockData, String> {
    let url = format!("https://query1.finance.yahoo.com/v8/finance/chart/{symbol}");
    let resp = http()?
        .get(&url)
        .header("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36")
        .header("Accept", "application/json")
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let v: Value = resp.json().await.map_err(|e| e.to_string())?;
    let meta = &v["chart"]["result"][0]["meta"];
    let price = meta["regularMarketPrice"].as_f64().unwrap_or(0.0);
    if price == 0.0 {
        return Err(format!("invalid price for {symbol}"));
    }
    let mut prev = meta["previousClose"].as_f64().unwrap_or(0.0);
    if prev == 0.0 {
        prev = meta["chartPreviousClose"].as_f64().unwrap_or(0.0);
    }
    let change = price - prev;
    let change_percent = if prev != 0.0 { change / prev * 100.0 } else { 0.0 };
    Ok(format_stock(symbol, price, change, change_percent))
}

async fn fetch_finnhub(symbol: &str, api_key: &str) -> Result<StockData, String> {
    let url = format!("https://finnhub.io/api/v1/quote?symbol={symbol}&token={api_key}");
    let v: Value = http()?.get(&url).send().await.map_err(|e| e.to_string())?.json().await.map_err(|e| e.to_string())?;
    let c = v["c"].as_f64().unwrap_or(0.0);
    if c == 0.0 {
        return Err(format!("no data for {symbol}"));
    }
    Ok(format_stock(symbol, c, v["d"].as_f64().unwrap_or(0.0), v["dp"].as_f64().unwrap_or(0.0)))
}

async fn fetch_alphavantage(symbol: &str, api_key: &str) -> Result<StockData, String> {
    let url = format!("https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol={symbol}&apikey={api_key}");
    let v: Value = http()?.get(&url).send().await.map_err(|e| e.to_string())?.json().await.map_err(|e| e.to_string())?;
    let q = &v["Global Quote"];
    if q["01. symbol"].as_str().unwrap_or("").is_empty() {
        return Err(format!("no data for {symbol}"));
    }
    let price = q["05. price"].as_str().unwrap_or("0").parse::<f64>().unwrap_or(0.0);
    let change = q["09. change"].as_str().unwrap_or("0").parse::<f64>().unwrap_or(0.0);
    let cp = q["10. change percent"]
        .as_str()
        .unwrap_or("0")
        .trim_end_matches('%')
        .parse::<f64>()
        .unwrap_or(0.0);
    Ok(format_stock(symbol, price, change, cp))
}
